### PR TITLE
router: abstract node key behind ECDH interface

### PR DESCRIPTION
**Pull request by guggero**
As part of the preparatory work for https://github.com/lightningnetwork/lnd/issues/3929, we remove the node's private key from the `NewRouter` method and replace it with an interface that abstracts away the ECDH operation.

This is only necessary for the long-term node key. All other private keys are just temporary session keys.

### CODE REVIEW

File: crypto.go [hunk 1]
```diff
@@ -22,6 +22,52 @@ const (
...
+// wrapped private key and remote public key. The output returned will be the
+// sha256 of the resulting shared point serialized in compressed format. If k is
+// our private key, and P is the public key, we perform the following operation:
+//
+//  sx := k*P
+//  s := sha256(sx.SerializeCompressed())
+//
+// NOTE: This is part of the SingleKeyECDH interface.
+func (p *PrivKeyECDH) ECDH(pub *btcec.PublicKey) ([32]byte, error) {
+	s := &btcec.PublicKey{}
```
Comment (Roasbeef) [inline crypto.go:65]: Use the `generateSharedSecret` method here just so we have all the ECDH code in one place?

File: sphinx.go [hunk 2]
```diff
@@ -131,7 +133,12 @@ func generateSharedSecrets(paymentPath []*btcec.PublicKey,
 	// Within the loop each new triplet will be computed recursively based
 	// off of the blinding factor of the last hop.
 	lastEphemeralPubKey := sessionKey.PubKey()
-	hopSharedSecrets[0] = generateSharedSecret(paymentPath[0], sessionKey)
+	hopSharedSecrets[0], err = generateSharedSecret(
```
Comment (Roasbeef) [inline sphinx.go:135]: No other errors in this function as is atm, so perhaps we should remove the error declaration above and just assign+declare the variable here? (walrus operator: `:=`)

File: crypto.go [hunk 3]
```diff
@@ -204,13 +204,10 @@ func (r *Router) generateSharedSecret(dhKey *btcec.PublicKey) (Hash256, error) {
...
 // single SHA256 invocation.  The resulting value is the shared secret.
-func generateSharedSecret(pub *btcec.PublicKey, priv *btcec.PrivateKey) (Hash256,
+func generateSharedSecret(pub *btcec.PublicKey, priv SingleKeyECDH) (Hash256,
 	error) {
 
-	s := &btcec.PublicKey{}
-	s.X, s.Y = btcec.S256().ScalarMult(pub.X, pub.Y, priv.D.Bytes())
-
-	return sha256.Sum256(s.SerializeCompressed()), nil
+	return priv.ECDH(pub)
```
Comment (Roasbeef) [inline crypto.go:210]: 👍

**Comment by Roasbeef** [comment]
cc @joostjager (reviewers thing is messed up on this repo, can't assign ppl easily)

File: crypto.go (hunk 1 above)
Comment (guggero) [reply crypto.go:65]: A later commit will change the code of `generateSharedSecret` to just call this method instead. So the code will be _moved_ here in effect. I added a commit message to explain this.

**Comment by guggero** [comment]
I'm working on a PR that does all this in lnd as well (`brontide`, `htlcswitch`, `netann` and `watchtower`). You mean that, right @Roasbeef? Or are there further changes that should be done in this repo?

Do you think the naming is OK? I'm going to use the same `SingleKeyECDH` thing in lnd too.

File: crypto.go (hunk 3 above)
Comment (joostjager) [inline crypto.go:210]: Looks like this function can be removed completely

File: crypto.go [hunk 4]
```diff
@@ -22,6 +22,52 @@ const (
 // the output of a SHA256 hash.
 type Hash256 [sha256.Size]byte
 
+// SingleKeyECDH is an abstraction interface that hides the implementation of an
+// ECDH operation against a specific private key. We use this abstraction for
+// the long term keys which we eventually want to be able to keep in a hardware
+// wallet or HSM.
```
[the remark this answers is not in the document]
Comment (guggero) [reply crypto.go:28]: I have no idea, to be honest. HW wallets are my blind spot. But I imagine implementing that is pretty easy as it's not much different from signing a message.

File: crypto.go [hunk 5]
```diff
@@ -22,6 +22,52 @@ const (
...
+type SingleKeyECDH interface {
+	// PubKey returns the public key of the wrapped, "local" private key.
```
[the remark this answers is not in the document]
Comment (guggero) [reply crypto.go]: Okay, adjusted the comment(s).

File: crypto.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (joostjager) [reply crypto.go:28]: @Roasbeef do they exist?

File: crypto.go [hunk 6]
```diff
@@ -22,6 +22,55 @@ const (
 // the output of a SHA256 hash.
 type Hash256 [sha256.Size]byte
 
+// SingleKeyECDH is an abstraction interface that hides the implementation of an
```
Comment (joostjager) [inline crypto.go:25]: Wondering still if `SingleKeyECDH` is the right name. It also suggests some fancy algorithm in which you exchange keys with yourself. Thinking about alternatives. Maybe `ExternalECDH` or just `Signer` (even though it isn't really signing)

File: crypto.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (Roasbeef) [reply crypto.go:28]: Ledger and trezor have support in their APIs which can be leveraged by custom applications. With this though, I think we're more so targeting remote/restricted signers which can emulate these calls by using the existing `lnd` gRPC interface and sub-servers, or something else entirely.

**Comment by Roasbeef** [comment]
> You mean that, right @Roasbeef? Or are there further changes that should be done in this repo?

Yeah, to propagate these changes into the `lnd` sphere. 

No strong feelings w.r.t naming on my end.

**Pull request merged**